llvm/test/Transforms/FunctionSpecialization/function-specialization-track-ptr-return.ll

```diff
@@ -0,0 +1,147 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --include-generated-funcs --version 5
+; RUN: opt -passes="ipsccp<func-spec>" -force-specialization \
+; RUN:     -funcspec-max-iters=3 -S < %s | FileCheck %s
+
+; RUN: opt -passes="ipsccp<func-spec>" -force-specialization          \
+; RUN:     -funcspec-for-literal-constant=false -funcspec-max-iters=3 \
+; RUN:     -S < %s | FileCheck %s --check-prefix=NOLIT
+
+@global_true = constant i1 true
+@global_false = constant i1 false
+
+define i64 @main() {
+entry:
+  %binop1 = call ptr @select_binop(ptr @global_true)
+  %binop2 = call ptr @select_binop(ptr @global_false)
+
+  %c1 = call i64 @compute(ptr %binop1)
+  %c2 = call i64 @compute(ptr %binop2)
+  %add = add i64 %c1, %c2
+  ret i64 %add
+}
+
+define ptr @select_binop(ptr %flag) {
+  %flag.val = load i1, ptr %flag
+  %binop = select i1 %flag.val, ptr @plus, ptr @minus
+  ret ptr %binop
+}
+
+define internal i64 @compute(ptr %binop) {
+entry:
+  %res = call i64 %binop(i64 1, i64 1)
+  ret i64 %res
+}
+
+define internal i64 @plus(i64 %x) {
+entry:
+  %sum = add i64 %x, 1
+  ret i64 %sum
+}
+
+define internal i64 @minus(i64 %x) {
+entry:
+  %diff = sub i64 %x, 1
+  ret i64 %diff
+}
+; CHECK-LABEL: define i64 @main() {
+; CHECK-NEXT:  [[ENTRY:.*:]]
+; CHECK-NEXT:    [[BINOP1:%.*]] = call ptr @select_binop.specialized.1(ptr @global_true)
+; CHECK-NEXT:    [[BINOP2:%.*]] = call ptr @select_binop.specialized.2(ptr @global_false)
+; CHECK-NEXT:    [[C1:%.*]] = call i64 @compute.specialized.3(ptr @plus)
+; CHECK-NEXT:    [[C2:%.*]] = call i64 @compute.specialized.4(ptr @minus)
+; CHECK-NEXT:    [[ADD:%.*]] = add i64 [[C1]], [[C2]]
```

Comment:
I am puzzled why the lattice value is not updated here. I'd think the entire main function can be folded to the constant value 2.

Reply:
Hmm yes good point - I'll see if I can figure out what is going on here.

Reply:
From a quick debug I noticed that plus and minus are not considered for specialization because they have not yet replaced the binop in the body of compute.specialized.3 and compute.specialized.4 (the Solver only knowns their lattice value), therefore findSpecializations does not find these callsites in the users list.

Reply:
Oh I see - I thought you meant that you would expect IPSCCP to find that plus and minus have constant return values, because they are only called once with a constant argument, but you meant you'd expect them to be specialized.

I think similarly IPSCCP doesn't handle propagating constants into function calls where the callee been found constant by the Solver. If you take the example IR:
```
define i64 @test() {
  %op = select i1 true, ptr @plus_test, ptr @minus_test
  %res = call i64 %op(i64 1)
  %sum = add i64 %res, 1
  ret i64 %sum
}
```
Running this through IPSCCP spits out:
```
`define i64 @test() {
  %res = call i64 @plus_test(i64 1)
  %sum = add i64 %res, 1
  ret i64 %sum
}

define internal i64 @plus_test(i64 %x) {
entry:
  %sum = add i64 %x, 1
  ret i64 %sum
}

define internal i64 @minus_test(i64 %x) {
entry:
  %diff = sub i64 %x, 1
  ret i64 %diff
}
`
define internal i64 @plus_test(i64 %x) {
entry:
  %sum = add i64 %x, 1
  ret i64 %sum
}

define internal i64 @minus_test(i64 %x) {
entry:
  %diff = sub i64 %x, 1
  ret i64 %diff
}
```

Reply:
I wonder if you could maybe teach `SCCPInstVisitor::visitCallBase` to handle indirect callees, although I'm not sure!

Reply:
I meant that SCCP is designed such that first it "solves" (analyzes the program, caches lattice values to its data structures) and then it replaces uses of values which are known to be constant. Things happen in this order by design. If you replaced the users earlier without invalidating the internal data structures (like resetLatticeValue does) then they would contain stale data and crash. I was fighting with such bugs when I merged FuncSpec with SCCP.

Reply:
I think it would be worth writing this finding in the header file of FuncSpec, in a new section under "ideas", perhaps named "limitations". Specifically not being able to consider specializations for callsites of indirect functions whose pointer operand has a lattice value that is known to be constant from previous iterations of FuncSpec. That is because SCCP has not yet replaced their uses.

Reply:
Thank you for clarifying :) I've added a comment to the header under `Limitations`, with the additional detail that the limitation also applies to indirect call-sites found constant by IPSCCP itself as well as previous FuncSpec iterations.